### PR TITLE
[RAPTOR-4183] refactoring: read input data as binary and pass into predictor

### PR DESCRIPTION
**Pull request by yakov-g**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary
Currently DRUM predictors (internals and hooks) receive data as either file or bytes.
Make it receiving bytes only. e.i. if file is detected, read it as binary and pass bytes further into predictor.

### CODE REVIEW

**Comment by yakov-g** [comment]
jarvis test cmrunner_integration

**Comment by yakov-g** [comment]
jarvis test cmrunner_integration

**Comment by yakov-g** [comment]
jarvis test cmrunner_integration

**Comment by yakov-g** [comment]
jarvis test cmrunner_integration

**Comment by yakov-g** [comment]
jarvis test cmrunner_integration

**Comment by yakov-g** [comment]
jarvis test cmrunner_integration

**Comment by yakov-g** [comment]
jarvis test cmrunner_integration

**Comment by yakov-g** [comment]
jarvis test cmrunner_integration

**Comment by yakov-g** [comment]
jarvis test cmrunner_integration

File: custom_model_runner/datarobot_drum/drum/language_predictors/java_predictor/java_predictor.py [hunk 1]
```diff
@@ -146,13 +146,10 @@ def supported_payload_formats(self):
...
     def predict(self, **kwargs):
-        input_filename = kwargs.get(StructuredDtoKeys.FILENAME)
-        input_binary_data = kwargs.get(StructuredDtoKeys.BINARY_DATA)
-        if input_binary_data is not None:
-            input_binary_data = input_binary_data.decode("utf-8")
+        input_binary_data = kwargs.get(StructuredDtoKeys.BINARY_DATA).decode("utf-8")
 
         start_predict = time.time()
-        out_csv = self._predictor_via_py4j.predict(input_filename, input_binary_data)
+        out_csv = self._predictor_via_py4j.predict(None, input_binary_data)
```
Comment (zohar-mizrahi) [inline custom_model_runner/datarobot_drum/drum/language_predictors/java_predictor/java_predictor.py:152]: Does this call with 'None' imply that the java interface was not revised to comply with the new changes?

File: custom_model_runner/datarobot_drum/drum/language_predictors/r_predictor/score.R [hunk 2]
```diff
@@ -210,19 +211,20 @@ outer_predict <- function(target_type, input_filename=NULL, binary_data=NULL, mo
...
-            data <- as.data.frame(as.matrix(readMM(input_filename)))
-        } else {
-            tmp = readChar(input_filename, file.info(input_filename)$size)
-            data <- read.csv(text=gsub("\r","", tmp, fixed=TRUE))
-        }
+    if (!isFALSE(read_input_data_hook)) {
+        data <- read_input_data_hook(binary_data)
+    } else if (!is.null(mimetype) && mimetype == "text/mtx") {
+        tmp_file_name <- tempfile()
+        f <- file(tmp_file_name, "wb")
```
Comment (zohar-mizrahi) [inline custom_model_runner/datarobot_drum/drum/language_predictors/r_predictor/score.R:218]: Isn't it possible in R to read the binary data directly from memory? Looks like a redundant step of writing to the disk.

File: custom_model_runner/datarobot_drum/drum/model_adapter.py [hunk 3]
```diff
@@ -348,22 +302,18 @@ def transform(self, model=None, **kwargs):
...
         pd.DataFrame
         """
-        input_filename = kwargs.get(StructuredDtoKeys.FILENAME)
         input_binary_data = kwargs.get(StructuredDtoKeys.BINARY_DATA)
         if self._custom_hooks.get(CustomHooks.READ_INPUT_DATA):
             try:
-                data = self._custom_hooks[CustomHooks.READ_INPUT_DATA](
-                    input_filename if input_filename is not None else input_binary_data
-                )
+                data = self._custom_hooks[CustomHooks.READ_INPUT_DATA](input_binary_data)
```
Comment (zohar-mizrahi) [inline custom_model_runner/datarobot_drum/drum/model_adapter.py:308]: Not fully aware for the context, but can you guarantee that the `self._custom_hooks` contains that entry?

File: custom_model_runner/datarobot_drum/drum/language_predictors/r_predictor/score.R (hunk 2 above)
Comment (yakov-g) [reply custom_model_runner/datarobot_drum/drum/language_predictors/r_predictor/score.R:218]: Unfortunately I'm not able to find smth like `BytesIO` in Python.

`readMM` can read from the R `connections` which all behave like a file: https://stat.ethz.ch/R-manual/R-devel/library/base/html/connections.html

File: custom_model_runner/datarobot_drum/drum/model_adapter.py (hunk 3 above)
Comment (yakov-g) [reply custom_model_runner/datarobot_drum/drum/model_adapter.py:308]: We check if before: ` if self._custom_hooks.get(CustomHooks.READ_INPUT_DATA):`

File: custom_model_runner/datarobot_drum/drum/language_predictors/java_predictor/java_predictor.py (hunk 1 above)
Comment (yakov-g) [reply custom_model_runner/datarobot_drum/drum/language_predictors/java_predictor/java_predictor.py:152]: Fixed java code

**Review by meltzg — APPROVED** [review]

**Review by zohar-mizrahi — APPROVED** [review]

**Comment by datarobot-jarvis** [comment]
:dollar: Total cost of executing tests in Jenkins: $**12.99** USD.

[What is this all about?](http://bit.ly/2BahK4V)

**Pull request merged**